test/testcontext/testcontext.go

```diff
@@ -0,0 +1,807 @@
+	}
+	ctx.Iterations = make([]*iteration.Iteration, ctx.info[kindIterations].numInstances)
+	for i := range ctx.Iterations {
+		desc := "Some description"
+		id := uuid.NewV4()
+		ctx.Iterations[i] = &iteration.Iteration{
+			ID:          id,
+			Name:        "iteration " + id.String(),
+			Description: &desc,
+		}
+		if !ctx.isolatedCreation {
+			ctx.Iterations[i].SpaceID = ctx.Spaces[0].ID
+		}
+		ctx.runDeferredCreationFunctions(i, kindIterations)
+		if ctx.isolatedCreation {
+			require.NotEqual(ctx.T, uuid.Nil, ctx.Iterations[i].SpaceID, "you must specify a space ID for each iteration")
+		}
+		err := iteration.NewIterationRepository(ctx.db).Create(ctx.ctx, ctx.Iterations[i])
+		require.Nil(ctx.T, err, "failed to create iteration: %+v", ctx.Iterations[i])
+	}
+}
+
+func makeAreas(ctx *TestContext) {
+	if ctx.info[kindAreas] == nil {
+		return
+	}
+	ctx.Areas = make([]*area.Area, ctx.info[kindAreas].numInstances)
+	for i := range ctx.Areas {
+		id := uuid.NewV4()
+		ctx.Areas[i] = &area.Area{
+			ID:   id,
+			Name: "area " + id.String(),
+		}
+		if !ctx.isolatedCreation {
+			ctx.Areas[i].SpaceID = ctx.Spaces[0].ID
+		}
+		ctx.runDeferredCreationFunctions(i, kindAreas)
+		if ctx.isolatedCreation {
+			require.NotEqual(ctx.T, uuid.Nil, ctx.Areas[i].SpaceID, "you must specify a space ID for each area")
+		}
+		err := area.NewAreaRepository(ctx.db).Create(ctx.ctx, ctx.Areas[i])
+		require.Nil(ctx.T, err, "failed to create area: %+v", ctx.Areas[i])
+	}
+}
+
+func makeCodebases(ctx *TestContext) {
+	if ctx.info[kindCodebases] == nil {
+		return
+	}
+	ctx.Codebases = make([]*codebase.Codebase, ctx.info[kindCodebases].numInstances)
+	for i := range ctx.Codebases {
+		id := uuid.NewV4()
+		stackID := "golang-default"
+		ctx.Codebases[i] = &codebase.Codebase{
+			ID:                id,
+			Type:              "git",
+			StackID:           &stackID,
+			LastUsedWorkspace: "my-used-last-workspace",
+			URL:               "git@github.com:fabric8-services/fabric8-wit.git",
+		}
+		if !ctx.isolatedCreation {
+			ctx.Codebases[i].SpaceID = ctx.Spaces[0].ID
+		}
+		ctx.runDeferredCreationFunctions(i, kindCodebases)
+		if ctx.isolatedCreation {
+			require.NotEqual(ctx.T, uuid.Nil, ctx.Codebases[i].SpaceID, "you must specify a space ID for each codebase")
+		}
+		err := codebase.NewCodebaseRepository(ctx.db).Create(ctx.ctx, ctx.Codebases[i])
+		require.Nil(ctx.T, err, "failed to create codebase: %+v", ctx.Codebases[i])
+	}
+}
+
+func makeWorkItemTypes(ctx *TestContext) {
+	if ctx.info[kindWorkItemTypes] == nil {
+		return
+	}
+	ctx.WorkItemTypes = make([]*workitem.WorkItemType, ctx.info[kindWorkItemTypes].numInstances)
+	for i := range ctx.WorkItemTypes {
+		desc := "this work item type was automatically generated"
+		id := uuid.NewV4()
+		ctx.WorkItemTypes[i] = &workitem.WorkItemType{
+			ID:          id,
+			Name:        "work item type " + id.String(),
+			Description: &desc,
+			Icon:        "fa-bug",
+			Fields: map[string]workitem.FieldDefinition{
```

Comment:
why do you need to redefine all these fields ? Don't we already have them in the database ?

Reply:
We only have them in the database as default WITs like systemBug or systemPlannerItem but when I need to define new items I think I have to do it this way. The only other option that comes to my mind is to load one of the other existing WITs from the database, change the ID and name and create a new WIT with that new ID and Name. I think I will do that. For now I more or less copied what was in the migration.go file.

Reply:
yes, or maybe some code from the migration can be externalized and reused ? (just a quick thought)